docs/templating.md

```diff
@@ -0,0 +1,188 @@
+<b>hub</b>
+<b>rip</b>
+```
+
+> **See Also:** [Mustache Manual](https://mustache.github.io/mustache.5.html) for more information on Sections.
+
+
+## Partials
+Along with sections, Mustache utilizes partials. Mustache partials can be thought of as a way to insert template snippets. The syntax for including a partial uses curley braces and an angle bracket {{> }}.
+
+For BIG-IP FAST, a partial definition must contain template text, i.e., define a template property
+```yaml
+definitions:
+  partialDef:
+    template: |
+      {{#useVar}}
+        {{var}}
+      {{/useVar}}
+  useVar:
+    type: boolean
+  template: |
+  {{> partialDef}}
+  {{> partialDef}}
+```
+
+Parameters:
+```yaml
+{
+    "useVar": true,
+    "var": "sample"
+}
+```
+
+Outputs:
+```yaml
+    sample
+    sample
+```
+
+> **See Also:** [Mustache Manual](https://mustache.github.io/mustache.5.html) for more information on Partials.
+
+
+# Template Data Files
+
+Sometimes it is desirable to keep a portion of a template in a separate file and include it into the template text.
+This can be done with parameters and the `dataFile` property:
+
+```javascript
+const fast = require('@f5devcentral/f5-fast-core');
+
+const templatesPath = '/path/to/templatesdir'; // directory containing example.data
+const dataProvider = new fast.FsDataProvider(templatesPath);
+const yamldata = `
+    definitions:
+        var:
+            dataFile: example
+    template: |
+        {{var}}
+`;
+
+fast.Template.loadYaml(yamldata, { dataProvider })
+    .then((template) => {
+        console.log(template.getParametersSchema());
+        console.log(template.render({virtual_port: 443});
+    });
+```
+The `FsDataProvider` will pick up on any files with the `.data` extension in the template set directory.
+When referencing the file in a template, use the filename (without the extension) as a key.
+
+Parameters with a `dataFile` property:
+
+* are removed from `required`
+* have their `default` set to the contents of the file
+* given a default `format` of `hidden`
+
+Additionally, the contents of the data file can be base64-encoded before being used as for `default` by setting the `toBase64` property to `true`:
+
+```yaml
+definitions:
+    var:
+        dataFile: example
+        toBase64: true
+    template: |
+        {{var}}
+```
+
+Similarly, if the data file is base64-encoded, it can be decoded using `fromBase64`.
+If both `toBase64` and `fromBase64` are set, then `toBase64` takes precedence.
+
+# Base64 Encoding
+Using the GUI, BIG-IP FAST has the ability to encode template parameters as base64, which becomes part of the template output (AS3 declaration).
+iRules are a common use case, however AS3 supports base64 for a wide range of objects.
+
+In the following example, base64var will display as editable plain text but render as base64-encoded text:
+```yaml
+contentType: application/json
+definitions:
+    base64var:
+        type: string
+        format: text
+        contentEncoding: base64
```

Comment:
toBase64 and fromBase64 are discussed at the end of this file, in the Template Data Files section. 
Should that be moved to the Base64 section -- which I have removed since it is specific to JSON Editor, while to/from base64 is data specific and might only belong where it is now?